[TySan][Clang] Add clang flag to use tysan outlined instrumentation a…

gbMattN opened the pull request:
…nd update docs

gbMattN committed:
[TySan][Clang] Add clang flag to use tysan outlined instrumentation and update docs

gbMattN committed:
Fix typo in Option.td

melver commented on clang/test/CodeGen/sanitize-type-outlined.cpp:
```diff
@@ -0,0 +1,23 @@
+// CHECK-OUTLINE-VERIFY: call{{.*}}@__tysan_instrument_mem_inst
+// CHECK-VERIFY: call{{.*}}@__tysan_instrument_mem_inst
+
+float alias(int *ptr){
```
Can you use CHECK-LABEL to avoid accidentally passing the test for some reason?

gbMattN commented on clang/test/CodeGen/sanitize-type-outlined.cpp:
```diff
@@ -0,0 +1,23 @@
+// RUN:     -fsanitize-type-verify-outlined-instrumentation \
+// RUN:     | FileCheck %s --check-prefixes=CHECK-VERIFY
+
+// CHECK-NO-OUTLINE-NOT: call{{.*}}@__tysan_instrument_mem_inst
```
Both __tysan_instrument_mem_inst and __tysan_instrument_with_shadow_update are functions which can be inserted when outlining is enabled, so either (perhaps better if both) can be used to check that the outlining of instrumentation is occurring. I made this test mimicking the minimal one ASan has but on second thought maybe I should make it a bit more verbose to properly check for all possible outlined functions

gbMattN committed:
Windows doesnt support tysan so unsupport in the test

fhahn reviewed (commented):
@gbMattN did you already get a chance to verify the outlined instrumentation matches via `-fsanitize-type-verify-outlined-instrumentation` on a larger set of workloads?

Are you aware of any cases where the outlined insturmentation performs worse than the inline instrumentation?

If not, I think it would be better to due some due diligence testing using `-fsanitize-type-verify-outlined-instrumentation` and flip the default, instead of adding the plumbing to make it convenient to use

gbMattN committed:
Switch outlining to default, remove verification focused flag changes

gbMattN committed:
Update docs

gbMattN commented:
Hi @fhahn. I've done some more exhaustive testing with @BStott6 and we're convinced that the instrumentation is behaving the same. I've removed the extra code I added to make the verify flag easier to use. I have not yet removed the verify option altogether; is this something you would be OK with?

gbMattN committed:
Fixup tests

gbMattN committed:
Re-add reset changes

gbMattN committed:
Add more reset changes

gbMattN committed:
review comments and fix that one outline test

gbMattN committed:
Merge branch 'main' into tysan-update-clang-for-outlining

fhahn commented on llvm/test/Instrumentation/TypeSanitizer/access-with-offset.ll:
```diff
@@ -1,7 +1,8 @@
-; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-globals all --version 6
-; RUN: opt -passes='tysan' -S %s | FileCheck %s
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
```
Should keep `--check-globals=all` to make sure chec lines for globals are (re)generated as needed?

fhahn reviewed (approved):
LGTM,thanks

gbMattN committed:
Update test to check for globals

gbMattN committed:
Actually regen the test
